src/Stiefel.jl

```diff
@@ -102,18 +103,27 @@ end
 
 compute the inverse retraction based on a singular value decomposition
 for two points `x`, `y` on the [`Stiefel`](@ref) manifold `M` and return
-the resulting tangent vector in `v`.
-The formula reads
+the resulting tangent vector in `v`. This follows the wolloing approach:
+From the Polar retraction we know that
 ````math
-v = ys-x$,
+    \operatorname{retr}_x^{-1}y = ys - q 
 ````
-where $s$ is the solution to the Lyapunov equation
+if such a symmetric positive definite $n\times n$ matrix exists. Since $ys-q$ is
+    also a tangent vector at $x$ we obtain
 ````math
 $x^{\mathrm{T}}ys + s(x^{\mathrm{T}}y)^{\mathrm{T}} + 2\mathrm{I}_k = 0.
 ````
+Setting $m=x^\mathrm{T}y$ this yields $(-m)s + s(-m^\mathrm{T}) + 2I = 0$,
+a continuous-time algebraic Riccati equation (arec). This follows Algorithm 2 in
+
+> T. Kaneko, S. Fiori, T. Tanaka: "Empirical Arithmetic Averaging over the
+> Compact Stiefel Manifold", IEEE Transactions on Signal Processing, 2013,
+> doi: [10.1109/TSP.2012.2226167](https://doi.org/10.1109/TSP.2012.2226167).
 """
 function inverse_retract!(::Stiefel{M,N,T}, v, x, y, ::PolarInverseRetraction) where {M,N,T}
-  v .= y*lyap(x'*y, -2*one(x'*x)) - x
+   # arec (A,Q,R) solves A'X + XA + XRX + Q = 0
```

Comment:
The arguments for `arec` are in a different order (A, R, Q). It's quite surprising that this solver is more numerically stable than `lyap` and `lyapc` (from `MatrixEquations`).

Reply:
And Travis reports errors from `MatrixEquations`

Reply:
Oh that's just a type on the comment. Indeed for the equation Q is 2I, so the call is correct and just the comment is wrong; thanks for noticing.

I have not tried `lyapc` – but `lyap` (from LinearAlgebra) was not that stable.

Reply:
That I do not understand: What's restriction `*` that combined with `1.0.0` only leaves `1.0.0`whcih I actually _have_ installed and which is available? On my Julia 1.2 the tests ran fine – I think I still have to learn how to add packages, such that tests don't fail.

Reply:
`MatrixEquations` requires at least Julia 1.2 (that's what they have in Project.toml)

Reply:
Oh, and they only test on Julia 1.2. Not even Julia 1.3. That's unfortunate.

Reply:
Hm, I don't have a solution for that, I was happy to find numerically a solution for the problem I was facing.

Reply:
We could improve `MatrixEquations` and make a PR but maybe we could do with `layp` from `LinearAlgebra` for now? And just have greater tolerances in tests?

Reply:
I am not sure what the PR would look like, but back to `lapis not that good of an idea, since it's more than tolerance – it's `NaN`s and `Inf`s that I fought most of the afternoon.

Reply:
Right, if it's `NaN`s and `Inf`s then it looks like we need `MatrixEquations`. I'll try to find some time to fix the library. For now you can convert the first argument of `arec` to `Array` (since `Transposed` seems to cause trouble with `hvcat` used in `MatrixEquations`). Travis on MacOS and Julia 1.2 also complains about some numerical inaccuracy of `project_tangent` for Stiefel.

Reply:
Cool, thanks; yes, `project_tangent` seems to be a tough one when it comes to accuracy, I already introduced a tolerance-variable for that, but ycould not find a way to get that more accurate. I will do the Array tomorrow somewhen.

Reply:
Some problems seem to come from type piracy in LinearOperators which could be fixed on master. There are still some accuracy issues. Finally, I've opened this issue in `ManifoldEquations.jl` https://github.com/andreasvarga/MatrixEquations.jl/issues/3 . I think we have to wait for now.

Reply:
I haven't completely checked the type piracy you mentioned – which types does that involve?

Reply:
Type piracy causes this error:
```
Type Array{Float32,2}: Error During Test at /home/travis/build/JuliaNLSolvers/Manifolds.jl/test/stiefel.jl:18
  Got exception outside of a @test
  MethodError: hvcat(::Tuple{Int64,Int64}, ::Adjoint{Float32,Array{Float32,2}}, ::Array{Float32,2}, ::Array{Float32,2}, ::Array{Float32,2}) is ambiguous. Candidates:
    hvcat(rows::Tuple{Vararg{Int64,N} where N}, xs::Union{AbstractArray{T,1}, AbstractArray{T,2}}...) where T in Base at abstractarray.jl:1634
    hvcat(rows::Tuple{Vararg{Int64,N} where N}, ops::Union{AbstractArray{T,2} where T, LinearOperators.AbstractLinearOperator}...) in LinearOperators at /home/travis/.julia/packages/LinearOperators/La2Ey/src/LinearOperators.jl:639
    hvcat(rows::Tuple{Vararg{Int64,N} where N}, xs::Union{AbstractArray{T,1}, AbstractArray{T,2}} where T...) in Base at abstractarray.jl:1633
  Possible fix, define
    hvcat(::Tuple{Vararg{Int64,N} where N}, ::Vararg{Union{AbstractArray{T,1}, AbstractArray{T,2}},N} where N)
  Stacktrace:
   [1] arec(::Adjoint{Float32,Array{Float32,2}}, ::Array{Float32,2}, ::Array{Float32,2}) at /home/travis/.julia/packages/MatrixEquations/MBfWx/src/riccati.jl:81
```
You can see that the method at `LinearOperators.jl:639` can be called with any matrices, without using any types defined by `LinearOperators.jl`. That's type piracy.